massa-deferred-calls/src/registry_changes.rs

```diff
@@ -0,0 +1,303 @@
+};
+use nom::{
+    error::{context, ContextError, ParseError},
+    multi::length_count,
+    sequence::tuple,
+    IResult, Parser,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    slot_changes::{
+        DeferredRegistrySlotChanges, DeferredRegistrySlotChangesDeserializer,
+        DeferredRegistrySlotChangesSerializer,
+    },
+    DeferredCall, DeferredRegistryGasChange,
+};
+use std::ops::Bound::Included;
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct DeferredRegistryChanges {
+    pub slots_change: BTreeMap<Slot, DeferredRegistrySlotChanges>,
+    pub total_gas: DeferredRegistryGasChange<u128>,
+}
+
+impl DeferredRegistryChanges {
+    pub fn merge(&mut self, other: DeferredRegistryChanges) {
+        for (slot, changes) in other.slots_change {
+            match self.slots_change.entry(slot) {
+                std::collections::btree_map::Entry::Occupied(mut entry) => {
+                    entry.get_mut().merge(changes);
+                }
+                std::collections::btree_map::Entry::Vacant(entry) => {
+                    entry.insert(changes);
+                }
+            }
+        }
+        match other.total_gas {
+            DeferredRegistryGasChange::Set(v) => self.total_gas = DeferredRegistryGasChange::Set(v),
+            DeferredRegistryGasChange::Keep => {}
+        }
+    }
+
+    pub fn delete_call(&mut self, target_slot: Slot, id: &DeferredCallId) {
+        self.slots_change
+            .entry(target_slot)
+            .or_default()
+            .delete_call(id)
+    }
+
+    pub fn set_call(&mut self, id: DeferredCallId, call: DeferredCall) {
+        self.slots_change
+            .entry(call.target_slot.clone())
+            .or_default()
+            .set_call(id, call);
+    }
+
+    pub fn get_call(&self, target_slot: &Slot, id: &DeferredCallId) -> Option<&DeferredCall> {
+        self.slots_change
+            .get(target_slot)
+            .and_then(|slot_changes| slot_changes.get_call(id))
+    }
+
+    pub fn get_slot_gas(&self, target_slot: &Slot) -> Option<u64> {
+        self.slots_change
+            .get(target_slot)
+            .and_then(|slot_changes| slot_changes.get_gas())
+    }
+
+    pub fn set_slot_gas(&mut self, target_slot: Slot, gas: u64) {
+        self.slots_change
+            .entry(target_slot)
+            .or_default()
+            .set_gas(gas);
+    }
+
+    pub fn set_slot_base_fee(&mut self, target_slot: Slot, base_fee: Amount) {
+        self.slots_change
+            .entry(target_slot)
+            .or_default()
+            .set_base_fee(base_fee);
+    }
+
+    pub fn get_slot_base_fee(&self, target_slot: &Slot) -> Option<Amount> {
+        self.slots_change
+            .get(target_slot)
+            .and_then(|slot_changes| slot_changes.get_base_fee())
+    }
+
+    pub fn set_total_gas(&mut self, gas: u128) {
+        self.total_gas = DeferredRegistryGasChange::Set(gas);
+    }
+
```

Comment:
there is no set_cancelled ?

Reply:
in Serializer ?

We have it in `speculative_deferred_calls` (`cancel_call()`), we set cancelled to true and then call `set_call()` of `DeferredCallRegistryChanges`

Reply:
https://github.com/massalabs/massa/blob/f3331b56b324ac3b59289e89ab593804e9710e31/massa-execution-worker/src/speculative_deferred_calls.rs#L293